### ISSUE TITLE
[52] SLIP: Self-supervision meets Language-Image Pre-training

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#80]
[paper](https://arxiv.org/pdf/2112.12750.pdf)
[code](https://github.com/facebookresearch/SLIP)

다음 이미지에서 볼 수 있듯이, (YFCC 15M 에서 SLIP pretrained)
image only self-supervision과 image-text supervision 을 동시에 주는 것은 성능향상에 도움이 된다.
![image](https://user-images.githubusercontent.com/16400591/147723083-5209d7d5-26ce-4174-b78a-3c98494e319f.png)

논문 주장:
```
Contrastive Image self-supervision 은
CLIP과 같은 Language-Image SSL 에도 도움이 된다.
```

# INTRO
YFCC 100M 데이터는 CLIP 에서 썼었음.
물론 CLIP의 curated 400M image 가 훨씬 성능이 좋았지만... 이거 써도 성능이 엄청 나쁘진 않았음
이게 15년도에 나온 셋인데, 사람들이 Imagenet-21K 를 더 많이 씀
YFCC 100M 은 curation 도 안되서 시작하기엔 부담스러운게 사실임.

YFCC 100M 데이터를 CLIP 방식으로 curation 해서 YFCC15M 을 만들고,
여러가지 실험 진행함
![image](https://user-images.githubusercontent.com/16400591/147730461-30240ed3-6e81-45b1-93a3-9b336c8748a2.png)

2가지 uncurated data도 pretrain에 활용해봐서 제안하는 방법의 효용성 보여줌
Conceptual Captions 3M (CC3M)
Conceptual Captions 12M (CC12M)

# SLIP Framework
그냥 2가지 loss 같이 줌
2가지 loss 사이의 lambda 값은 1

```python
# fi, ft: image, text encoders
# hi, ht: CLIP image, text projectors
# hs: SimCLR projector
# c: SimCLR loss scale
def forward(img, text):
    xi, x1, x2 = crop(img), aug(img), aug(img)
    yt = tokenize(text)
    wi, w1, w2 = fi(xi, x1, x2)
    wt = ft(yt)
    z1, z2 = hs(w1), hs(w2) # SSL embed: N x C2
    zi, zt = hi(wi), ht(wt) # CLIP embed: N x C1
    loss = c * simclr(z1, z2) + clip(zi, zt)
    return loss


# s: learnable log logit scale
def clip(zi, zt):
    zi, zt = normalize(zi, zt)
    label = range(N)
    logit = exp(s) * zi @ zt.T
    li = CrossEntropy(logit, label)
    lt = CrossEntropy(logit.T, label)
    loss = (li + lt) / 2
    return loss


# tau: softmax temperature
def simclr(z1, z2):
    z1, z2 = normalize(z1, z2)
    label = range(N)
    mask = eye(N) * 1e9
    logit = z1 @ z2.T
    logit1 = z1 @ z1.T - mask
    logit2 = z2 @ z2.T - mask
    logit1 = cat(logit, logit1)
    logit2 = cat(logit.T, logit2)
    l1 = CrossEntropy(logit1 / tau)
    l2 = CrossEntropy(logit2 / tau)
    loss = (l1 + l2) / 2
    return loss
```

추가로 Improved training procedure 를 개발함
다른 모델도 fair comparison 하기위해서 전부 같은 procedure 적용.
모델들 성능 전부 올랐음. new training procedure 유의미함.
1. random resize-crop 50%~100% (저자들은 global cropping 이라고 naming 함)
2. view 마다 적용하는 aug 는 moco v3 에서 하듯이 함

(1) 이 유의미했다고 함

# Results
## ImageNet
ImageNet 결과들. (with improved training techniques)
YFCC pretrain은 여기서 imagenet에 구린 성능을 가져다 주는 걸로 보임
![image](https://user-images.githubusercontent.com/16400591/147732391-a0788ac4-b869-41ee-b0e5-3d4038522b92.png)

## CLIP vs SLIP (ImageNet)
YFCC 15M 에서 25epoch 학습 결과.
SLIP 은 LINEAR 나 FINETUNE 에서 모두 뛰어난 수치를 보여줌
![image](https://user-images.githubusercontent.com/16400591/147732458-031660c3-fdc7-465f-913f-d00a1664e405.png)
![image](https://user-images.githubusercontent.com/16400591/147732693-e83891f2-3b6b-43c1-9e48-c3323928e8be.png)

## Scale Exp
100epoch 학습시키면 약간 더 좋아짐
![image](https://user-images.githubusercontent.com/16400591/147732738-1060af20-70fd-4e86-9d7e-cc6b4e2060a8.png)

## CLIP vs SLIP (various sets)
![image](https://user-images.githubusercontent.com/16400591/147732777-7034cb37-a009-4ae9-a19e-6b2dbb433174.png)

## Other uncurated dataset pretrain 은??
CC3M, CC12M pretrain으로 ImageNet 에 transfer 해봐도 효과적
![image](https://user-images.githubusercontent.com/16400591/147732929-fb1a6968-d762-4030-a613-f34ccf185fcf.png)

## 다른 contrastive learning 기법과의 궁합
ViT-B/16 에 이것저것 붙여보니까 SimCLR 이 젤 나아 보임
![image](https://user-images.githubusercontent.com/16400591/147732878-b30e93a3-0125-4b0e-b8d5-c6fe7bd6f83a.png)
